### PR TITLE
Added learning phase to callbacks (#2297)

### LINKED ISSUE #2297: TensorBoard with Histograms crashes with Dropout Layers
The most recent version of Keras crashes after the first epoch when a TensorBoard callback is added that records histograms.

Failing Example Code (`examples/mnist_mlp.py` with added callback) and Crash Log:
https://gist.github.com/jdoerrie/713a2d53e14b01470aa664494186d7c7

Judging by the output the issue is related to `keras_learning_phase` not being fed. The code works normally when `histogram_freq` is set to 0 or the Dropout layers are removed.

- Operating System: Mac OS X 10.11.3
- Python: 3.5.1
- Keras: 1.0.0 (Master Checkout)
- Tensorflow: 0.7.1
- Running on CPU.

### PR DESCRIPTION
**Pull request by tboquet**
This PR closes #2297.

The `_LEARNING_PHASE` tensor was not added to the `feed_dict` passed to the TensorFlow session.
The `model.validation_data` also had too much information (the target validation data).
I chose to append a 0 to the list instead of cutting it to be concise.

I'm not sure about how we could retrieve the info in a cleaner way so let me know if it's ok!

If we add some more metrics it should be taken in account but for now it does the job and it's explicit.
This could be changed to use the information already contained in the list.

### CODE REVIEW

File: keras/callbacks.py [hunk 1]
```diff
@@ -462,13 +462,16 @@ def _set_model(self, model):
 
     def on_epoch_end(self, epoch, logs={}):
         import tensorflow as tf
+	import keras.backend.tensorflow_backend as KTF
```
Comment (fchollet) [inline keras/callbacks.py]: You can simply use `K.learning_phase()` and remove this import.

**Comment by fchollet** [comment]
> The model.validation_data also had too much information (the target validation data).

Can you clarify?

File: keras/callbacks.py (hunk 1 above)
Comment (tboquet) [inline keras/callbacks.py]: Oops, sure. It's also breaking the tests.

**Comment by tboquet** [comment]
If I'm not mistaking, for the visualizations of the weights, we only need the `model.inputs` corresponding arrays. We also have the sample weights and the outputs in the `model.validation_data` list.
When the list is zipped with the `model.inputs` the tensor returned by `K.learning_phase()` is coupled with the wrong array (the first output).

**Comment by tboquet** [comment]
Really, sorry for the multiple builds, I did the changes on a remote server and my emacs config got mixed up. I will squash the commits when the build will pass.

**Comment by tboquet** [comment]
Ok, the callback is not working for Graph models and it breaks a lot of other tests too :(. 
The learning phase tensor is not part of the computation graph when using `Graph` models and it's the remaining problem.

Should we add it in the graph somewhere so it's possible to have backward compatibility when using the Tensorboard Callback with a `Graph` model?

The test for the TensorBoard callback is skipped for python 3.4, should we reactivate it?

**Comment by fchollet** [comment]
> The test for the TensorBoard callback is skipped for python 3.4, should we reactivate it?
 Hide all checks

Yes, sounds reasonable.

> The learning phase tensor is not part of the computation graph when using Graph models and it's the remaining problem.

Why? Can you clarify? I'm not following.

**Comment by tboquet** [comment]
My mistake, it didn't work with `Sequential` either.
I just figured out the `uses_learning_phase` attribute existed.
It should work now :)... I'm still discovering the mechanics of the new tools.
 
I also reactivated all the callback tests, they seemed to be disabled during the big migration.

**Comment by tboquet** [comment]
There is the traditionnal download error in one of the build. Let me know if I have to change something or if I can squash the commits!

**Comment by tboquet** [comment]
Trying to relaunch the tests...

**Comment by fchollet** [comment]
LGTM. Thanks!
